# QRT ENS Data Challenge 2023 - Benchmark (english)

This notebook details the benchmark construction for the challenge - it may also help participants to start the competition.

## Libraries

In [11]:
import numpy as np
import pandas as pd
from scipy.stats import spearmanr
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

## Loading data

- `X_train` and `X_test` both have $35$ columns that represent the same explanatory variables but over different time periods. 

- `X_train` and `Y_train` share the same column `ID` - each row corresponds to a unique ID associated wwith a day and a country. 

- The target of this challenge `TARGET` in `Y_train` corresponds to the price change for daily futures contracts of 24H electricity baseload. 

- **You will notice some columns have missing values**.


In [12]:
# After downloading the X_train/X_test/Y_train .csv files in your working directory:

X_train = pd.read_csv('X_train.csv')
Y_train = pd.read_csv('Y_train.csv')
X_test = pd.read_csv('X_test.csv')

In [13]:
X_train.head()

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0  1054     206      FR        0.210099       -0.427458       -0.606523   
1  2049     501      FR       -0.022399       -1.003452       -0.022063   
2  1924     687      FR        1.395035        1.978665        1.021305   
3   297     720      DE       -0.983324       -0.849198       -0.839586   
4  1101     818      FR        0.143807       -0.617038       -0.924990   

   FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
0        0.606523            NaN       0.692860            NaN  ...   
1        0.022063      -0.573520      -1.130838       0.573520  ...   
2       -1.021305      -0.622021      -1.682587       0.622021  ...   
3        0.839586      -0.270870       0.563230       0.270870  ...   
4        0.924990            NaN       0.990324            NaN  ...   

   FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
0         -0.444661 -0.172680 -0.556356 -0.790823 -0.283160 -1.069070   
1         -1.183194 -1.240300 -0.770457  1.522331  0.828412  0.437419   
2          1.947273 -0.480700 -0.313338  0.431134  0.487608  0.684884   
3         -0.976974 -1.114838 -0.507570 -0.499409 -0.236249  0.350938   
4         -0.526267 -0.541465 -0.424550 -1.088158 -1.011560  0.614338   

    FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
0 -0.063404  0.339041  0.124552   -0.002445  
1  1.831241 -0.659091  0.047114   -0.490365  
2  0.114836  0.535974  0.743338    0.204952  
3 -0.417514  0.911652 -0.296168    1.073948  
4  0.729495  0.245109  1.526606    2.614378  

[5 rows x 35 columns]

In [14]:
Y_train.head()

ID    TARGET
0  1054  0.028313
1  2049 -0.112516
2  1924 -0.180840
3   297 -0.260356
4  1101 -0.071733

## Model and train score

The benchark for this challenge consists in a simple linear regression, after a light cleaning of the data: The missing (NaN) values are simply filled with 0's and the `COUNTRY` column is dropped - namely we used the same model for France and Germany. 

In [15]:
lr = LinearRegression()

X_train_clean = X_train.drop(['COUNTRY'], axis=1).fillna(0)
Y_train_clean = Y_train['TARGET']

lr.fit(X_train_clean, Y_train_clean)

output_train = lr.predict(X_train_clean)

def metric_train(output):

    return  spearmanr(output, Y_train_clean).correlation

print('Spearman correlation for the train set: {:.1f}%'.format(100 * metric_train(output_train) ))

Spearman correlation for the train set: 27.9%


The Spearman correlation obtained with our model on the train data set is about 27.9%.

NB: Electricity price variations can be quite volatile and this is why we have chosen the Spearman rank correlation as a robust metric for the challenge, instead of the more standard Pearson correlation.
- A coefficient of +1 indicates a perfect positive rank correlation, -1 indicates a perfect negative rank correlation, and 0 indicates no correlation.

## Generate the benchmark output

Next, we process the test set the same way as we did on the train set and predict using our linear model, while saving the predictions to a csv file satisfying the challenge output contraints.


In [16]:
X_test.head()

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
0  1115     241      FR        0.340083       -0.433604       -0.423521   
1  1202    1214      FR        0.803209        0.780411        0.601610   
2  1194    1047      FR        0.795540        0.721954        1.179158   
3  1084    1139      FR        0.172555       -0.723427       -0.044539   
4  1135     842      FR        0.949714        0.420236        0.617391   

   FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
0        0.423521       0.165333       0.519419      -0.165333  ...   
1       -0.601610       0.342802       0.555367      -0.342802  ...   
2       -1.179158       1.620928       0.666901      -1.620928  ...   
3        0.044539            NaN      -0.205276            NaN  ...   
4       -0.617391       0.608561      -0.240856      -0.608561  ...   

   FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
0         -0.222525 -0.513180 -0.182048 -0.982546 -0.876632  0.880491   
1          0.857739 -0.340595 -0.301094 -0.759816 -1.221443 -0.616617   
2          0.447967  0.796475 -0.367248  0.376055 -0.483363  0.865138   
3         -0.561295 -0.542606 -0.013291 -0.791119 -0.894309  0.239153   
4          0.503567 -0.230291 -0.609203 -0.744986 -1.196282  0.176557   

    FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
0  0.692242  0.569419 -0.029697   -0.929256  
1 -0.737496  0.251251  0.753646    0.664086  
2  0.120079 -1.485642 -0.326450   -0.349747  
3  0.457457 -0.746863  2.262654    0.642069  
4  0.312557 -2.219626 -0.509272   -0.488341  

[5 rows x 35 columns]

In [17]:
X_test_clean = X_test.drop(['COUNTRY'], axis=1).fillna(0)

Y_test_submission = X_test[['ID']].copy()
Y_test_submission['TARGET'] = lr.predict(X_test_clean)

Y_test_submission.to_csv('benchmark_qrt.csv', index=False)


After submitting this benchmark_qrt.csv file at https://challengedata.ens.fr, we obtain a public score of around 15.9 % - this is the score to beat!


## Tips and ideas for improvements 

- Thinking about the fundamental price drivers of electricity in each country may be useful. 

- The dataset is small, so you will need to be careful not to overfit your models. It will be useful to use good crossvalidation practices. 


# another test change

# EDA (Vivian)

- France 

In [60]:
fr_df = X_train[X_train['COUNTRY'] == 'FR']
de_df = X_train[X_train['COUNTRY'] == 'DE']

#make the dataframe descending by DAY_IDX
fr_df = fr_df.sort_values(by='DAY_ID', ascending=True)
de_df = de_df.sort_values(by='DAY_ID', ascending=True)



In [72]:
fr_df

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
742   1057       0      FR        0.048418       -0.465067       -0.079296   
1211  1179       1      FR        0.966997        1.222131        0.331356   
1408  1327       2      FR       -0.068972       -0.667390       -1.102015   
329   2016       3      FR       -0.134670       -0.834564       -1.051716   
512   2047       5      FR       -0.297850       -0.470371        0.144615   
...    ...     ...     ...             ...             ...             ...   
479   2039    1208      FR        0.046568       -0.709011        0.427976   
1451   972    1209      FR        1.109804        1.301061        1.565552   
1007  1987    1212      FR       -0.046100       -0.520506       -0.270515   
873   1688    1213      FR        1.407027        1.313254       -0.456126   
1437  1049    1215      FR        0.544810       -0.378158       -0.065993   

      FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
742         0.079296       0.000000       0.650577       0.000000  ...   
1211       -0.331356       0.405549       0.778627      -0.405549  ...   
1408        1.102015      -1.080403       0.256736       1.080403  ...   
329         1.051716      -1.881881      -0.612133       1.881881  ...   
512        -0.144615      -1.208286      -1.811403       1.208286  ...   
...              ...            ...            ...            ...  ...   
479        -0.427976      -0.578391      -1.678101       0.578391  ...   
1451       -1.565552       0.000000       0.000000       0.000000  ...   
1007        0.270515      -0.137917      -0.789824       0.137917  ...   
873         0.456126       0.333142      -0.069222      -0.333142  ...   
1437        0.065993       0.000000       0.257829       0.000000  ...   

      FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
742          -0.344706 -0.987968 -0.856204 -0.655528 -1.099226  0.279885   
1211          1.214288 -0.370990 -0.497520 -1.184018 -1.465608  1.349212   
1408         -0.540642 -0.344600 -0.372156 -0.623041 -0.926064 -0.148950   
329          -0.856321 -1.394561 -1.118297  0.244859  0.488650 -1.710888   
512          -0.552878 -1.374757 -0.790071  0.333817  0.021868 -0.062187   
...                ...       ...       ...       ...       ...       ...   
479          -0.666261 -1.224735  2.661142  0.472212  0.630211  1.065727   
1451          1.038585  0.457145 -0.202629 -0.395438 -0.463636 -0.251075   
1007         -0.589767 -0.039379 -0.237434  1.358211  1.123953 -0.707839   
873           0.868310  0.000000  0.000000  0.000000  0.000000  0.000000   
1437         -0.223383 -0.646163  0.738201 -0.903380 -0.882630 -0.247278   

       FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
742   0.583874 -1.071996 -0.074767    0.798003  
1211  0.231602  1.480313  0.931562    0.822047  
1408  0.641235  1.802550  1.140920    0.900434  
329  -0.951057  0.440121 -0.064550   -0.032756  
512   1.459745 -0.117977  0.550433    0.781870  
...        ...       ...       ...         ...  
479  -1.439105  0.291714  0.568479    0.395742  
1451 -0.387127 -0.411589  1.774694    1.038661  
1007 -0.308232  0.017778  0.072168   -0.160792  
873   0.000000 -0.103642 -0.633007    0.214805  
1437  0.755489  0.758698  0.078518    0.192945  

[851 rows x 35 columns]

In [73]:
de_df

ID  DAY_ID COUNTRY  DE_CONSUMPTION  FR_CONSUMPTION  DE_FR_EXCHANGE  \
1468  111       2      DE       -0.068972       -0.667390       -1.102015   
680   800       3      DE       -0.134670       -0.834564       -1.051716   
991   831       5      DE       -0.297850       -0.470371        0.144615   
1356  779       7      DE        0.057599       -0.625625        0.002239   
713   841       8      DE       -0.282744       -0.700027        1.309253   
...   ...     ...     ...             ...             ...             ...   
168   282    1202      DE       -1.182561       -1.259840       -1.201004   
1136  505    1207      DE        1.002463        0.516789        0.651648   
577   823    1208      DE        0.046568       -0.709011        0.427976   
986   771    1212      DE       -0.046100       -0.520506       -0.270515   
1465  472    1213      DE        1.407027        1.313254       -0.456126   

      FR_DE_EXCHANGE  DE_NET_EXPORT  FR_NET_EXPORT  DE_NET_IMPORT  ...  \
1468        1.102015      -1.080403       0.256736       1.080403  ...   
680         1.051716      -1.881881      -0.612133       1.881881  ...   
991        -0.144615      -1.208286      -1.811403       1.208286  ...   
1356       -0.002239      -0.676226      -0.745182       0.676226  ...   
713        -1.309253       0.248085      -2.226840      -0.248085  ...   
...              ...            ...            ...            ...  ...   
168         1.201004      -1.616737       0.619501       1.616737  ...   
1136       -0.651648       0.728817      -0.820640      -0.728817  ...   
577        -0.427976      -0.578391      -1.678101       0.578391  ...   
986         0.270515      -0.137917      -0.789824       0.137917  ...   
1465        0.456126       0.333142      -0.069222      -0.333142  ...   

      FR_RESIDUAL_LOAD   DE_RAIN   FR_RAIN   DE_WIND   FR_WIND   DE_TEMP  \
1468         -0.540642 -0.344600 -0.372156 -0.623041 -0.926064 -0.148950   
680          -0.856321 -1.394561 -1.118297  0.244859  0.488650 -1.710888   
991          -0.552878 -1.374757 -0.790071  0.333817  0.021868 -0.062187   
1356         -0.806379 -0.531862 -0.663419  0.098940  1.367421 -0.812164   
713          -0.791016  1.522720 -0.307449  1.690497  1.584210 -0.441943   
...                ...       ...       ...       ...       ...       ...   
168          -1.359708 -0.085663 -0.997403 -0.595651 -0.777719  0.039450   
1136          0.236243       NaN       NaN       NaN       NaN       NaN   
577          -0.666261 -1.224735  2.661142  0.472212  0.630211  1.065727   
986          -0.589767 -0.039379 -0.237434  1.358211  1.123953 -0.707839   
1465          0.868310       NaN       NaN       NaN       NaN       NaN   

       FR_TEMP   GAS_RET  COAL_RET  CARBON_RET  
1468  0.641235  1.802550  1.140920    0.900434  
680  -0.951057  0.440121 -0.064550   -0.032756  
991   1.459745 -0.117977  0.550433    0.781870  
1356  0.954384 -0.379980  0.518459   -0.034642  
713  -0.382423  1.557325  0.146634    2.329924  
...        ...       ...       ...         ...  
168   1.411276  0.184767  0.257420    0.507079  
1136       NaN  0.099209  1.750872    0.646905  
577  -1.439105  0.291714  0.568479    0.395742  
986  -0.308232  0.017778  0.072168   -0.160792  
1465       NaN -0.103642 -0.633007    0.214805  

[643 rows x 35 columns]

In [63]:
empty_values = fr_df.isnull().sum()
print(empty_values)

ID                    0
DAY_ID                0
COUNTRY               0
DE_CONSUMPTION        0
FR_CONSUMPTION        0
DE_FR_EXCHANGE       25
FR_DE_EXCHANGE       25
DE_NET_EXPORT       124
FR_NET_EXPORT        70
DE_NET_IMPORT       124
FR_NET_IMPORT        70
DE_GAS                0
FR_GAS                0
DE_COAL               0
FR_COAL               0
DE_HYDRO              0
FR_HYDRO              0
DE_NUCLEAR            0
FR_NUCLEAR            0
DE_SOLAR              0
FR_SOLAR              0
DE_WINDPOW            0
FR_WINDPOW            0
DE_LIGNITE            0
DE_RESIDUAL_LOAD      0
FR_RESIDUAL_LOAD      0
DE_RAIN              47
FR_RAIN              47
DE_WIND              47
FR_WIND              47
DE_TEMP              47
FR_TEMP              47
GAS_RET               0
COAL_RET              0
CARBON_RET            0
dtype: int64


In [69]:
fr_df = fr_df.fillna(0)


In [70]:
empty_values = fr_df.isnull().sum()
print(empty_values)

ID                  0
DAY_ID              0
COUNTRY             0
DE_CONSUMPTION      0
FR_CONSUMPTION      0
DE_FR_EXCHANGE      0
FR_DE_EXCHANGE      0
DE_NET_EXPORT       0
FR_NET_EXPORT       0
DE_NET_IMPORT       0
FR_NET_IMPORT       0
DE_GAS              0
FR_GAS              0
DE_COAL             0
FR_COAL             0
DE_HYDRO            0
FR_HYDRO            0
DE_NUCLEAR          0
FR_NUCLEAR          0
DE_SOLAR            0
FR_SOLAR            0
DE_WINDPOW          0
FR_WINDPOW          0
DE_LIGNITE          0
DE_RESIDUAL_LOAD    0
FR_RESIDUAL_LOAD    0
DE_RAIN             0
FR_RAIN             0
DE_WIND             0
FR_WIND             0
DE_TEMP             0
FR_TEMP             0
GAS_RET             0
COAL_RET            0
CARBON_RET          0
dtype: int64


In [64]:
empty_values = de_df.isnull().sum()
print(empty_values)

ID                   0
DAY_ID               0
COUNTRY              0
DE_CONSUMPTION       0
FR_CONSUMPTION       0
DE_FR_EXCHANGE       0
FR_DE_EXCHANGE       0
DE_NET_EXPORT        0
FR_NET_EXPORT        0
DE_NET_IMPORT        0
FR_NET_IMPORT        0
DE_GAS               0
FR_GAS               0
DE_COAL              0
FR_COAL              0
DE_HYDRO             0
FR_HYDRO             0
DE_NUCLEAR           0
FR_NUCLEAR           0
DE_SOLAR             0
FR_SOLAR             0
DE_WINDPOW           0
FR_WINDPOW           0
DE_LIGNITE           0
DE_RESIDUAL_LOAD     0
FR_RESIDUAL_LOAD     0
DE_RAIN             47
FR_RAIN             47
DE_WIND             47
FR_WIND             47
DE_TEMP             47
FR_TEMP             47
GAS_RET              0
COAL_RET             0
CARBON_RET           0
dtype: int64
